### PR TITLE
DO-NOT-MERGE Feed plugin sanitation

### PR DESCRIPTION
**Pull request by millette**
I couln't make the feed plugin work until I finally understood the reason. I had an object in ```siteMetadata``` instead of a string or a number.

It was failing with this error:
```
error Plugin gatsby-plugin-feed returned an error


  TypeError: Cannot convert object to primitive value
  
  - xml.js:177 resolve
    [gatsby-starter-location-github]/[xml]/lib/xml.js:177:34
  
  - xml.js:191 
    [gatsby-starter-location-github]/[xml]/lib/xml.js:191:42
  
  - Array.forEach
  
  - xml.js:184 resolve
    [gatsby-starter-location-github]/[xml]/lib/xml.js:184:24
  
  - xml.js:191 
    [gatsby-starter-location-github]/[xml]/lib/xml.js:191:42
  
  - Array.forEach
  
  - xml.js:184 resolve
    [gatsby-starter-location-github]/[xml]/lib/xml.js:184:24
  
  - xml.js:47 add
    [gatsby-starter-location-github]/[xml]/lib/xml.js:47:24
  
  - xml.js:89 xml
    [gatsby-starter-location-github]/[xml]/lib/xml.js:89:9
  
  - index.js:185 RSS.xml
    [gatsby-starter-location-github]/[rss]/lib/index.js:185:13
  
  - gatsby-node.js:151 _loop$
    [gatsby-starter-location-github]/[gatsby-plugin-feed]/gatsby-node.js:151    :73
```  

All I had to do to fix it was to remove that object from my feed query configuration and thought maybe this could help someone else. The patch simply removes any value under ```siteMetadata``` that isn't a string or a number, and also converts numbers to strings (for use in the rss feed).

Feel free to merge or not. If not, maybe we could add a bit of documentation about that.

### CODE REVIEW

**Comment by secretfader** [comment]
The issue raised here concerns me for a few reasons. One, I use objects and arrays in `siteMetadata`, and often store global configuration there to be queried by Gatsby later on. I imagine other users do as well.

Second, even if we were to sanitize query results to either a number or string, this concern *isn't limited to the feed plugin.* If we're concerned about users placing objects in the `siteMetadata`, it's my perspective that we should disallow the behavior entirely (possibly in Gatsby core) not just in one plugin.

I'm certainly open to discussing further, but I feel strongly that we'd need to resolve the issues above before moving forward.

**Comment by millette** [comment]
Should I make it an issue (I'm totally fine with that) and let you move your comment there? By closing the PR, it sort of dissappears.

**Comment by pieh** [comment]
I don't think there is any problem with keeping deep nested objects in `siteMetadata` - only problem here was that xml serializer didn't handle that in feed plugin, correct?

**Comment by pieh** [comment]
Could you add some tests for this? Recently some tests were added to feed plugin (https://github.com/gatsbyjs/gatsby/pull/8580/files), so you would have to merge our master in

**Comment by millette** [comment]
I found it a bit hard to deal with tests and the multipackage repo, but I'll give it a shot within a day or two. I'll also make sure it [lints](https://circleci.com/gh/gatsbyjs/gatsby/5297).

**Comment by secretfader** [comment]
As I said previously, I'm against this change (for a myriad of issues, but primarily reasons mentioned above). If there's an issue with the underlying XML generator, shouldn't we replace *that package*, rather than patching up the perceived flaws in a Gatsby plugin, with a breaking change, no less?

None of the issues I mentioned above have been resolved. I'm fairly sure this will break several builds I'm responsible for, and I'd appreciate it if the discussion didn't continue as if I never objected.

**Comment by pieh** [comment]
What is the breaking change here? This PR doesn't affect things outside of feed plugin. It adds sanitizing query results to something that current serializer can handle. Of course adding support for things that currently can't be handled would be better, but this seem like better short term action than allowing builds to break with unhelpful error message?

**Comment by pieh** [comment]
@secretfader Can you elaborate in what way this change would break things? It seems either me or you are missing something here and would be good to clarify things.

**Comment by secretfader** [comment]
@pieh You may want to go re-read my comments above, because my objections were clear. 

I'm the original author of this plugin :wave:. As such, I'm also one of the heaviest users of it. I routinely insert objects into `siteMetadata`, to be later queried by the feed plugin, and others as well. This PR removes a supported feature and introduces a breaking change to correct user error.

It's unfortunate that some users have experienced issues due to the complexities of this specific plugin (see past issues for reference), and I've tried many different strategies to simplify the API. Every attempt has ended up failing because this plugin is widely used, across many cases, and difficult to simplify without breaking many builds across the community.

Writing and querying for objects (again, which I routinely do), is a supported feature. If the underlying RSS generator has issues parsing configuration *other than strings and numbers*, it's the users' job to translate their custom configuration into a format it can accept using the `setup` callback.

**Comment by pieh** [comment]
@secretfader I did re-read your comments (again) but until your last one it still wasn't clear to me so thank You for clarifying things.

> If the underlying RSS generator has issues parsing configuration _other than strings and numbers_, it's the users' job to translate their custom configuration into a format it can accept using the `setup` callback.

If users need custom `setup` function, we could probably improve DX by catching those scenarios and providing more meaningful feedback (like what should they adjust in gatsby-config). If we could run setup and serialize part before we even interact with `rss` object we could see if any of feed items would contain plain objects / supported fields by `rss` package and we could warn users about it - is this correct?

> Every attempt has ended up failing because this plugin is widely used, across many cases, and difficult to simplify without breaking many builds across the community.

If you have ideas how plugin can be improved that need breaking changes for users we can always bump major version to do this - it's not ideal but if it needs to be done, then it needs to be done.

**Comment by secretfader** [comment]
@pieh That idea might be worth a try.

> If we could run setup and serialize part before we even intera, often strange behavior of their own.ct with rss object we could see if any of feed items would contain plain objects / supported fields by rss package and we could warn users about it...

That avenue is certainly an option. Though I don't know if we, as a static site generator, should be in the business of enforcing APIs of existing packages. Primarily because many RSS/Atom generators in the Node ecosystem often have their own, unique idiosyncrasies.

tldr: If someone wants to write a better implementation, I'm generally in favor of improving DX for this plugin. However, ensuring that a user properly configured the underlying XML generator may be more trouble than we want to deal with. I expect that Gatsby will eventually need to implement it's own feed generator, if we're hoping to improve DX past a certain point.
